### ISSUE TITLE
sphinx.util.inspect.getargspec tries to access an attribute in func.keywords and to delete an item in func_defaults when it has a partial function, when both could be None

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1069]
In `sphinx.util.inspect.getargspec` there is this code:

```
#!python

        if type(func) is partial:
            parts = len(func.args), func.keywords.keys()

```

`func.keywords` may be None when a partial function has no keywords; in such cases, accessing `keys` raises AttributeError.

Still in `getargspec`, a similar thing happens with `func.func_defaults`:

```
#!python

                try:
                    del func_defaults[i]
                except IndexError:
                    pass

```

`func_defaults` can be None as well, when a function has no default argument values; in such cases this will raise a TypeError.

These issues also affect autodoc, which uses `getargspec` to get function signatures.

I have attached a patch fixing these issues.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1069
- Originally reported by: Artur Gaspar
- Originally created at: 2013-01-08T15:00:43.584

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-09 14:59:34+00:00_

Closes #1069: Fix autodoc signature formatting of "partial" functions without kwargs (patch by Artur Gaspar).

→ <<cset 61e951d2ed83933e39f76aada92b546db3036c26>>

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-09 15:00:43+00:00_

Thanks
